tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,524 @@
+        //                continue;
+        //            }
+        //            assert!(!authkeystext.contains(&*TEST_KEYS[key]));
+        //        }
+    }
+
+    fn add_key_check_results(ssh_dir: &str, aks: AuthorizedKeys, keys: Vec<AuthorizedKeyEntry>, pubkeyname: &str, testname: &str, assert_keys: Vec<&str>) {
+        for key in &keys {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = *key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                aks.write().expect("failed to update authorized keys directory");
+                aks.sync().expect("failed to update authorized keys");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(authkeyone.exists());
+
+                assert_has_keys(ssh_dir, assert_keys.clone(), true);
+            }
+        }
+    }
+
+    fn open_authorized_keys(ssh_dir: &str) -> AuthorizedKeys {
+        let ssh_dir: PathBuf = PathBuf::from(ssh_dir);
+        let unameosstr = users::get_current_username().unwrap_or_default();
+        let unamestr = unameosstr.to_str().unwrap_or_default();
+        let user =
+            users::get_user_by_name(&unamestr).ok_or_else(|| format!("failed to find user with name '{}'", unamestr)).unwrap_or(users::User::new(UID_CORE as uid_t, USERNAME_CORE, GID_CORE as gid_t));
+
+        AuthorizedKeys::open(user, true, Some(ssh_dir)).expect(format!("failed to open authorized_keys directory for user '{}'", unamestr).as_str())
+    }
+
+    // A wrapper for adding an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like "valid1", "bad"
+    // testname: key name given as cmdline args, like "one", "two"
+    // is_force: whether to force adding a key via "--add-force"
+    // is_replace: whether to adding a key by replacing an existing one, unless "--no-replace".
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn add_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str, is_force: bool, is_replace: bool, assert_keys: Vec<&str>) -> Result<(), ErrorKind> {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        let res = aks.add_keys(testname, keys.clone(), is_replace, is_force);
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
```

Comment:
In `main.rs` I see that the writing/syncing is only done in for the `KeysDisabled` and `KeysExist` error, otherwise not.

Reply:
Done